### PR TITLE
R RNN API fixes and Optimizer clip gradient on NDArray

### PR DESCRIPTION
**Pull request by jeremiedb**
## Description ##
Rework of the RNN API: 
- Change the input to be of shape `seq_len X batch.size`, this will ease the use of existing arrayIter on NLP models. 
- Add support for inference on explicit unrolled models (used when training on CPU rather than the `symbol.RNN`)
- Inference support for both one-to-one (language model, time-series) and seq-to-one (classification). 

Optimizers:
For all optimizers, gradient clipping is now performed by `mx.nd.clip`, avoiding to back and forth with R array resulting in some speedup. 

Following comments on lack of exmaples supporting the R package, exemples have been added: 
https://jeremiedb.github.io/mxnet_R_bucketing/index.html
If there's interest, idea to is add other examples, in the spirit of the gluon.mxnet.io notebooks

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by jeremiedb** [comment]
@thirdwing : I think these changes should bring a broader usability of the RNN API.

**Comment by thirdwing** [comment]
I will look into this now. Sorry for the late reply. Too busy with the thesis.

**Comment by thirdwing** [comment]
Can you also update the code in https://github.com/dmlc/dmlc.github.io/blob/master/_posts/2017-10-11-rnn-bucket-mxnet-R.md ?

**Comment by jeremiedb** [comment]
Yes, just created the pull requests on dmlc.github.io and web-data repos.
